### ISSUE TITLE
Should overloads with implicit conversion be preferred over instantiating?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20539]
One thing discussed in earlier versions of the effort in PR #20528 (and in particular, PR #19959) was if a generic overload should be chosen rather than a non-generic overload that requires implicit conversion. For example:

``` chapel
proc g(arg)       { writeln("in generic g"); }

proc g(arg: real) { writeln("in real g"); }

g(1);

```

The rules to date would call have `g(1)` call the `real` version of `g` rather than the generic one. PR #20528 preserves that behavior. One rationale for it is that programmers might have an idea when they create the overloads that they're making a concrete functions for some cases and then a generic version for erroneous cases. I saw a fair amount of test or module code that had that assumption.

*However* this is very different from what C++ does in an equivalent situation:

``` c++
// prog.cpp
#include <stdio.h>

template<typename T>
static void g(T arg) { printf("in generic g\n"); }

void g(double arg)   { printf("in real g\n"); }

int main() {
  int x = 1;
  g(x);

  return 0;
}
```

```
$ g++ prog.cpp && ./a.out
in generic g
```

Similarly, the C# version calls the generic function in a similar situation:

``` c#
// prog.cs

// to run this on linux
// sudo apt-get install mono-complete

// mcs -out:prog.exe prog.cs
// mono prog.exe

using System;

class MyTest {
  static void g<T>(T arg)   { Console.WriteLine("in generic g"); }
  static void g(double arg) { Console.WriteLine("in real g"); } 

  static void Main() {
    int int_ = 0;

    g(int_);
  }
}
```

```
$ mcs -out:prog.exe prog.cs
$ mono prog.exe
in generic g
```

Should this element of the Chapel language be changed?

**Comment by benharsh** [comment]
> I saw a fair amount of test or module code that had that assumption.

Were there any cases you thought were particularly notable?

> However this is very different from what C++ does in an equivalent situation

Do you know what rationale they provide?

-----

Is class inheritance a relevant concern?
```chpl

class Parent {
  var x : int;
}

class Child : Parent {
  var y : string;
}

proc g(arg) { writeln("generic"); }
proc g(arg:borrowed Parent) { writeln("concrete"); }

proc main() {
  var c = new Child();
  g(c);
}
```

**Comment by bradcray** [comment]
I feel as though my intuition is that the generic is a better match (because it's an exact match), whereas the coercion requires some sort of type change.  Like Ben, I'd be curious for cases that suggest doing the opposite (though I'm generally disinclined to depart from C++ in this respect).  I know that error overloads may be one such case, but that feels like a place where a "last resort" / "error case" feature is the better approach (where my expectation would be that the coercion would win in the presence of a pragma).

**Comment by mppf** [comment]
> > I saw a fair amount of test or module code that had that assumption.
> 
> Were there any cases you thought were particularly notable?

I don't remember, but I had a lot of other changes in flight at the same time. I think a good next step is to do focused experiment to answer this question.

> 
> > However this is very different from what C++ does in an equivalent situation
> 
> Do you know what rationale they provide?

I don't know, no. 

> Is class inheritance a relevant concern?

The example you showed prints out `concrete` which might be surprising.

**Comment by mppf** [comment]
I'm writing about some cases that rely on this behavior today.

`_cond_test` -- https://github.com/chapel-lang/chapel/blob/0ca1166e18947fea8820b67ce7d8833ebb88cf08/modules/internal/ChapelBase.chpl#L737-L765

`myInt += myInt32` -- https://github.com/chapel-lang/chapel/blob/0ca1166e18947fea8820b67ce7d8833ebb88cf08/modules/internal/ChapelBase.chpl#L1858-L1874

`_validRankChangeArg` https://github.com/chapel-lang/chapel/blob/0ca1166e18947fea8820b67ce7d8833ebb88cf08/modules/internal/ChapelArray.chpl#L1997-L1999

**Comment by mppf** [comment]
I just noticed that the behavior here is different when using a generic formal type for the generic function:

``` chapel
proc h(arg: integral) { writeln("in generic h"); }
proc h(arg: real)     { writeln("in real h"); }
h(1);
```

```
in generic h
```

(Vs the case in the original post, with `proc g(arg)` which ends up calling the `proc g(arg: real)` version).

I think that the "concrete is better than dtAny" was added in 3b20f318a63ceea6680a917e58375ff3bcea79f1, and then a types-must-match check was added in 78f57ae8ea1cdd2c297432bcd71411da5321765b. Anyway, it has been this way (and I think, inconsistent) for a very long time. I was able to verify the inconsistency in 1.17.
